[MRG+2] Fix trustworthiness custom metric

wdevazelhes opened the pull request:
#### Reference Issue
Fixes #9736 

#### What does this implement/fix? Explain your changes.
- It changes the way to specify a precomputed metric into a more standard way: metric='precomputed' instead of a flag precomputed=False/True. 
- In the meantime it allows to use a custom metric for the input space 
- However, it warns the user if they want to use another metric than euclidean and precomputed, since this is not standard.
- If the specified metric is 'euclidean', it keeps the 'squared' parameter in pairwise_distances from the original code.
- It adds a deprecation warning for the users who will still use the previous flag 'precomputed'

Commit:
Allow custom metric for sklearn.manifold.t_sne.trustworthiness #9736

Commit:
Fix the position of the deprecation in docstring.

massich commented on sklearn/manifold/t_sne.py:
```diff
@@ -413,15 +415,36 @@ def trustworthiness(X, X_embedded, n_neighbors=5, precomputed=False):
         Trustworthiness of the low-dimensional embedding.
     """
     if precomputed:
+        warnings.warn("The flag 'precomputed' has been deprecated in version"
```
Check [scikit-learn contributing guide](http://scikit-learn.org/stable/developers/contributing.html#deprecation) to see howto deprecate the `precomputed` attribute using a decorator.

massich replied:
and `precomputed` is not an attribute but a parameter. (my bad. sorry)

massich reviewed (commented):
Thanks for the PR.

massich commented:
Can you add a test ? Thx. 
You can use `test_grid_search_fit_params_deprecation()` as example (see [here] (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/model_selection/tests/test_search.py#L212))

Commit:
- Adds tests for deprecation and warnings
- Correct typos in output messages

Commit:
Fix almost equal error to be compatible with numpy 1.8.0

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -413,15 +415,36 @@ def trustworthiness(X, X_embedded, n_neighbors=5, precomputed=False):
+    metric : string, or callable, optional, default 'euclidean'
+        Which metric to use for computing pairwise distances between samples
+        from the original input space. If metric is 'precomputed', X must be a
+        matrix of pairwise distances. Otherwise, see the documentation of
```
Can you say "pairwise distances or squared distances" (either will work as this function only uses their rank, not their value).

jnothman commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -286,11 +288,48 @@ def test_preserve_trustworthiness_approximately_with_precomputed_distances():
+    # If a 'metric' different from 'precomputed' is specified, but the flag
+    # 'precomputed' is set to True, the flag overwrites the parameter 'metric'
+    # and so the 'precomputed' metric will be used. Indeed, in
+    # http://scikit-learn.org/stable/developers/contributing.html#deprecation,
```
I don't think we need this reminder... But yes, this is the what we tend to do.

jnothman commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -286,11 +288,48 @@ def test_preserve_trustworthiness_approximately_with_precomputed_distances():
+    # and so the 'precomputed' metric will be used. Indeed, in
+    # http://scikit-learn.org/stable/developers/contributing.html#deprecation,
+    # the old parameter overwrites the new parameter.
+    assert_raises(Exception, assert_warns, DeprecationWarning,
```
Usually we would not test for raising something as broad as Exception. Indeed, usually we'd check for a particular error message.

jnothman commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -286,11 +288,48 @@ def test_preserve_trustworthiness_approximately_with_precomputed_distances():
+    # Other metrics than 'euclidean' and 'precomputed' are unusual and must
+    # raise a warning
+    X = np.arange(100).reshape(50, 2)
+    assert_warns(RuntimeWarning, trustworthiness, X, X, metric='manhattan')
```
I actually strongly believe we should not raise a warning here or make any statement that this is abnormal. Even if deriving the manifold from an unusual metric is weird, checking that neighborhoods are maintained in the embedded space is precisely how we determine whether the learnt manifold was appropriate.

jnothman reviewed (commented):
Also, please add an entry to doc/whats_new

Commit:
Update commit after comments:

- indicate that one can use pairwise squared distances instead of distances
- remove useless reminder
- remove warning if metric used is unusual
- update doc/whats_new

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -100,3 +109,12 @@ Linear, kernelized and related models
+  ``metric`` should be used with any compatible metric including
+  'precomputed', in which case the input matrix ``X`` should be a matrix of
+  pairwise distances or squared distances. :issue:`9736` by
+  :user:`Joel Nothman <jnothman>`.
```
This should be your name and the issue number should be 9775

massich replied:
@jnothman 9775 is the id of the PR not the issue. Do we refer to the issue or to the PR? [here](https://github.com/scikit-learn/scikit-learn/pull/9059/files#diff-3433b590ca8611f80d7687b51a869383R40) we are referring to the issue. Is there a defined criteria? If so, I think we should add it in the contributing guide [here](http://scikit-learn.org/stable/developers/contributing.html#coding-guidelines) and be part of the reviewing checklist discussed in #9653

jnothman commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -286,11 +288,36 @@ def test_preserve_trustworthiness_approximately_with_precomputed_distances():
+    # 'precomputed'
+    X = np.array([[1, 1], [1, 0], [2, 2]])
+    assert_almost_equal(trustworthiness(X, X, n_neighbors=1, metric='cosine'),
+                        0.66, decimal=1)
```
.66 appears drawn from thin air. Instead benchmark against the corresponding precomputed distances

jnothman commented:
usually the pr. feel free to contribute it to contributing guide

jnothman commented:
having the pr there makes it easier for us to see how well the changelog
covers the set of commits in a release

Commit:
Modifications after comments:
- change user name and issue number in whats_new
- improve test of metric different than euclidean and precomputed

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -413,15 +415,33 @@ def trustworthiness(X, X_embedded, n_neighbors=5, precomputed=False):
+                      "0.20 and will be removed in 0.22. See 'metric' "
+                      "parameter instead.", DeprecationWarning)
+        metric = 'precomputed'
+    if metric == 'precomputed':
```
I think pairwise_distances handled precomputed, so we don't need a special case here

wdevazelhes replied:
Thanks, I will change this

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -413,15 +415,33 @@ def trustworthiness(X, X_embedded, n_neighbors=5, precomputed=False):
+    if metric == 'precomputed':
         dist_X = X
+    elif metric == 'euclidean':
+        dist_X = pairwise_distances(X, metric='euclidean', squared=True)
```
I similarly don't think it's essential to specially handle Euclidean, but at least it saves some computation

wdevazelhes replied:
I will change this too

jnothman reviewed (commented):
Otherwise LGTM

Commit:
Simplify calling pairwise_distances function

- Also, the error thrown when passing an array of vectors instead of pairwise distances
when 'precomputed' is used is not the same, so I changed the IndexError into ValueError
in test_trustworthiness_precomputed_deprecation

GaelVaroquaux commented:
LGTM. +1 for merge.

Will merge once the merge conflicts are resolved!

Commit:
Merge branch 'master' into 9736-trustworthiness-custom-metric

# Conflicts:
#	doc/whats_new/v0.20.rst
#	sklearn/manifold/t_sne.py

Commit:
Make random points in trustworthiness to avoid problems with equidistant KNeighbors

wdevazelhes commented:
Thanks ! I just resolved the conflicts
(Note that the AppVeyor build fail seems to be related to a connection timeout)

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -224,6 +224,15 @@ Linear, kernelized and related models
+Decomposition, manifold learning and clustering
+
+- Deprecate ``precomputed`` parameter in function
+  :func:`manifold.t_sne.trustworthiness`. Instead, the new parameter
```
If we have a separate entry for enhancement it should be written as such. For example, trustworthiness now accepts a metric other than Euclidean.

glemaitre committed:
Update v0.20.rst

glemaitre committed:
nitpicks

glemaitre commented:
@wdevazelhes I made the what's new changes and a change NOTE to FIXME. I think that we have more chance to find it during maintenance.

Merging now. Thanks!!!

wdevazelhes commented:
Thanks !
